File: python3/call_c/test_correctness.py
"""Correctness tests: Python3 -> C via MetaFFI

Tests C guest module entities accessible through MetaFFI.
Fail-fast: every assertion is exact (or epsilon-justified for floats).
"""

import pytest
import metaffi

T = metaffi.MetaFFITypes
ti = metaffi.metaffi_type_info


# ============================================================================
# Core functions (C guest)
# ============================================================================

class TestCoreFunctions:

    def test_no_op(self, c_module):
        fn = c_module.load_entity("callable=xcall_c_no_op", None, None)
        fn()
        del fn

    def test_hello_world(self, c_module):
        fn = c_module.load_entity("callable=xcall_c_hello_world", None,
            [ti(T.metaffi_string8_type)])
        result = fn()
        assert result == "Hello World from C"
        del fn

    def test_returns_an_error(self, c_module):
        """C guest returns int32 (-1) instead of throwing."""
        fn = c_module.load_entity("callable=xcall_c_returns_an_error", None,
            [ti(T.metaffi_int32_type)])
        result = fn()
        assert result == -1, f"returns_an_error: got {result}, want -1"
        del fn

    def test_div_integers(self, c_module):
        fn = c_module.load_entity("callable=xcall_c_div_integers",
            [ti(T.metaffi_int64_type), ti(T.metaffi_int64_type)],
            [ti(T.metaffi_float64_type)])
        result = fn(10, 2)
        assert abs(result - 5.0) < 1e-10, f"div_integers(10,2) = {result}, want 5.0"
        del fn


# ============================================================================
# State: counter
# ============================================================================

class TestState:

    def test_counter_operations(self, c_module):
        """set_counter, get_counter, inc_counter round-trip."""
        get_fn = c_module.load_entity("callable=xcall_c_get_counter", None,
            [ti(T.metaffi_int64_type)])
        set_fn = c_module.load_entity("callable=xcall_c_set_counter",
            [ti(T.metaffi_int64_type)], None)
        inc_fn = c_module.load_entity("callable=xcall_c_inc_counter",
            [ti(T.metaffi_int64_type)],
            [ti(T.metaffi_int64_type)])

        # Reset counter to 0
        set_fn(0)
        assert get_fn() == 0

        # Increment by 5
        new_val = inc_fn(5)
        assert new_val == 5, f"inc_counter(5) = {new_val}, want 5"
        assert get_fn() == 5

        # Reset
        set_fn(0)
        assert get_fn() == 0

        del get_fn, set_fn, inc_fn

File: java/call_c/src/test/java/TestCorrectness.java
import api.MetaFFIModule;
import api.MetaFFIRuntime;
import metaffi.api.accessor.Caller;
import metaffi.api.accessor.MetaFFITypeInfo;
import metaffi.api.accessor.MetaFFITypeInfo.MetaFFITypes;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Correctness tests: Java host -> C guest via MetaFFI.
 *
 * Tests C guest module entities accessible through MetaFFI.
 * C guest uses the "cpp" runtime (xllr.cpp handles both C and C++).
 * Fail-fast: every assertion is exact (or epsilon-justified for floats).
 */
public class TestCorrectness
{
	private static MetaFFIRuntime runtime;
	private static MetaFFIModule cModule;

	@BeforeClass
	public static void setUp()
	{
		String metaffiHome = System.getenv("METAFFI_HOME");
		assertNotNull("METAFFI_HOME must be set", metaffiHome);

		String sourceRoot = System.getenv("METAFFI_SOURCE_ROOT");
		assertNotNull("METAFFI_SOURCE_ROOT must be set", sourceRoot);

		// C guest uses the "cpp" runtime
		runtime = new MetaFFIRuntime("cpp");
		runtime.loadRuntimePlugin();

		String modulePath = sourceRoot.replace('\\', '/') +
			"/sdk/test_modules/guest_modules/c/test_bin/" + getCGuestModuleFilename();
		cModule = runtime.loadModule(modulePath);
		assertNotNull("Failed to load C guest module", cModule);
	}

	private static String getCGuestModuleFilename()
	{
		String os = System.getProperty("os.name", "").toLowerCase();
		if (os.contains("win"))
		{
			return "c_guest_module.dll";
		}
		if (os.contains("mac"))
		{
			return "c_guest_module.dylib";
		}
		return "c_guest_module.so";
	}

	@AfterClass
	public static void tearDown()
	{
		try
		{
			if (runtime != null)
			{
				runtime.releaseRuntimePlugin();
			}
		}
		finally
		{
			cModule = null;
			runtime = null;
		}
	}

	// ---- Helper methods ----

	private static MetaFFITypeInfo t(MetaFFITypes type)
	{
		return new MetaFFITypeInfo(type);
	}

	private Caller load(String entity, MetaFFITypeInfo[] params, MetaFFITypeInfo[] retvals)
	{
		Caller c = cModule.load(entity, params, retvals);
		assertNotNull("Failed to load entity: " + entity, c);
		return c;
	}

	// ========================================================================
	// Core functions (C guest)
	// ========================================================================

	@Test
	public void testNoOp()
	{
		Caller fn = load("callable=xcall_c_no_op", null, null);
		fn.call();
	}

	@Test
	public void testHelloWorld()
	{
		Caller fn = load("callable=xcall_c_hello_world", null,
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIString8)});
		Object[] result = fn.call();
		assertNotNull(result);
		assertEquals("Hello World from C", result[0]);
	}

	@Test
	public void testReturnsAnError()
	{
		// C guest returns int32 (-1) instead of throwing
		Caller fn = load("callable=xcall_c_returns_an_error", null,
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIInt32)});
		Object[] result = fn.call();
		assertNotNull(result);
		assertEquals(-1, ((Number) result[0]).intValue());
	}

	@Test
	public void testDivIntegers()
	{
		Caller fn = load("callable=xcall_c_div_integers",
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIInt64), t(MetaFFITypes.MetaFFIInt64)},
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIFloat64)});
		Object[] result = fn.call(10L, 2L);
		assertNotNull(result);
		assertEquals(5.0, (Double) result[0], 1e-10);
	}

	// ========================================================================
	// State: counter
	// ========================================================================

	@Test
	public void testCounter()
	{
		Caller getCtr = load("callable=xcall_c_get_counter", null,
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIInt64)});
		Caller setCtr = load("callable=xcall_c_set_counter",
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIInt64)}, null);
		Caller incCtr = load("callable=xcall_c_inc_counter",
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIInt64)},
			new MetaFFITypeInfo[]{t(MetaFFITypes.MetaFFIInt64)});

		// Reset to known value
		setCtr.call(0L);
		Object[] getResult = getCtr.call();
		assertEquals(0L, getResult[0]);

		// Increment by 5
		Object[] incResult = incCtr.call(5L);
		assertEquals(5L, incResult[0]);

		// Verify
		getResult = getCtr.call();
		assertEquals(5L, getResult[0]);

		// Reset
		setCtr.call(0L);
	}
}
